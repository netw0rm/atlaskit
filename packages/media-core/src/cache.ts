/**
 * TODO: Replace other LRUCache usage by this
 */
import {LRUCache} from 'lru-fast';

const CACHE_SIZE = 200;
const cache = new LRUCache(CACHE_SIZE);

export const set = (key: string = '', value: any) : any => {
  cache.set(key, value);

  return get(key);
};

export const get = (key: string = '') : any => {
  return cache.get(key);
};

export const getWithDefault = (key: string = '', value: Function) : any => {
  if (get(key)) { return get(key); }

  return set(key, value());
};

export default {
  set,
  get,
  getWithDefault
};
