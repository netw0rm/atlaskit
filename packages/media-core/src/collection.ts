import {MediaItemType} from '..';

export interface RemoteCollection {
  id: string;
  apiConfig: {
    clientId: string;
    token: string;
    baseURL: string;
  };
}

export interface MediaCollectionItem {
  id: string;
  mediaItemType: MediaItemType;
}

export interface MediaCollection {
  id: string;
  items: Array<MediaCollectionItem>;
}
