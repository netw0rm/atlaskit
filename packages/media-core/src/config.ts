export type JwtToken = string;

export type JwtTokenProvider = (collectionName?: string) => Promise<JwtToken>;

export type MediaApiConfig = {
  serviceHost: string,
  tokenProvider: JwtTokenProvider
};
