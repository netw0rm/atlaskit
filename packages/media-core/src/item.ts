import { MediaType } from './mediaTypes';
export type MediaItemType = 'file' | 'link';

export type MediaItem = FileItem | LinkItem;

export type MediaItemDetails = FileDetails | LinkDetails | UrlPreview;

export interface FileItem {
  type: 'file';
  details: FileDetails;
}

export type FileProcessingStatus = 'pending' | 'running' | 'succeeded' | 'failed';

export interface FileDetails {
  id?: string;
  name?: string;
  size?: number;
  mimeType?: string;
  mediaType?: MediaType;
  creationDate?: number; // timestamp in milliseconds from EPOCH
  processingStatus?: FileProcessingStatus;
  artifacts?: Object;
}

export interface LinkItem {
  type: 'link';
  details: LinkDetails;
}

export interface LinkDetails extends UrlPreview {
  id: string;
}

export interface UrlPreview {
  type: string;
  url: string;
  title: string;
  description?: string;
  site?: string;
  author?: UrlAuthorDetails;
  date?: number;
  resources?: Resources;
}

export interface UrlAuthorDetails {
  name?: string;
  url?: string;
}

export interface Resources {
  icon?: Resource;
  thumbnail?: Resource;
  image?: Resource;
  file?: Resource;
  player?: Resource;
  app?: App;
}

export interface Resource {
  url?: string;
  type?: string;
  width?: number;
  height?: number;
  length?: number;
  html?: string;
}

export interface App {
  title: AppTitle;
  description?: AppDescription;
  details?: AppDetail[];
  context?: AppContext;
  preview?: {url: string}; // an image URL
  background?: {url: string}; // an image URL
}

export interface AppTitle {
  text: string;
  user?: AppUser;
  lozenge?: AppLozenge;
}

export interface AppDescription {
  title?: string; // the bolded bit
  text: string;
}

export interface AppIcon {
  url: string;    // an image URL
  label: string;  // accessibility text e.g. tooltip or voiceover
}

export interface AppUser {
  icon: AppIcon;
}

export interface AppBadge {
  value: number;
  max?: number;
  theme?: 'default' | 'dark';
  appearance?: 'default' | 'primary' | 'important' | 'added' | 'removed';
}

export interface AppLozenge {
  text: string;
  bold?: boolean;
  appearance?: 'default' | 'success' | 'removed' | 'inprogress' | 'new' | 'moved';
}

export interface AppDetail {
  title?: string;
  icon?: AppIcon;
  badge?: AppBadge;
  lozenge?: AppLozenge;
  users?: AppUser[];
  text?: string;
}

export interface AppContext {
  text: string;
  icon?: AppIcon;
}
