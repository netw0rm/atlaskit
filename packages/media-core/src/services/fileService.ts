import createRequest from './util/createRequest';
import {FileItem, MediaApiConfig, MediaItem} from '../';
import {LRUCache} from 'lru-fast';

export interface FileService {
  getFileItem(fileId: string, clientId: string, collection?: string): Promise<FileItem>;
}

export class MediaFileService implements FileService {

  constructor(private config: MediaApiConfig, private cache: LRUCache<string, MediaItem>) {

  }

  getFileItem(fileId: string, clientId: string, collectionName?: string): Promise<FileItem> {
    const request = createRequest({
      config: this.config,
      clientId: clientId,
      collectionName: collectionName
    });

    const cacheKey = [fileId, 'file'].join('-');
    const cachedValue = this.cache.get(cacheKey);

    if (cachedValue) {
      return Promise.resolve(cachedValue);
    } else {
      return request({url: `/file/${fileId}`})
        .then(json => json.data)
        .then(fileDetails => {
          const fileItem =  <FileItem> {
            type: 'file',
            details: {
              id: fileDetails.id,
              mediaType: fileDetails.mediaType,
              mimeType: fileDetails.mimeType,
              name: fileDetails.name,
              processingStatus: fileDetails.processingStatus,
              size: fileDetails.size,
              artifacts: fileDetails.artifacts
            }
          };
          if (fileDetails.processingStatus === 'succeeded') {
            this.cache.set(cacheKey, fileItem);
          }
          return fileItem;
        });
      ;
    }
  }
}
