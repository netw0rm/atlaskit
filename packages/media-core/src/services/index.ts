export {DataUriService, DataUri} from './dataUriService';
