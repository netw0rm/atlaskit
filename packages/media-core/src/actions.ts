/* tslint:disable:variable-name */
import {MediaItem, FileDetails} from './item';

export type CardEventHandler = (item?: MediaItem, event?: Event) => void;
export type ListEventHandler = (item?: MediaItem, siblings?: Array<FileDetails>, event?: Event) => void;

export enum CardActionType {
  click, delete, download, retry, custom
}

export interface CardAction {
  label?: string;
  type?: CardActionType;
  handler: CardEventHandler;
}

export interface ListAction {
  label?: string;
  type?: CardActionType;
  handler: ListEventHandler;
}

export type CardActionCreator = (eventHandler: CardEventHandler) => CardAction;
export type ListActionCreator = (eventHandler: ListEventHandler) => ListAction;

export const CardClick: CardActionCreator = (eventHander: CardEventHandler) => {
  return {
    type: CardActionType.click,
    handler: eventHander
  };
};

export const CardDelete: CardActionCreator = (eventHander: CardEventHandler) => {
  return {
    label: 'Delete',
    type: CardActionType.delete,
    handler: eventHander
  };
};

export const ListCardClick: ListActionCreator = (eventHander: ListEventHandler) => {
  return {
    type: CardActionType.click,
    handler: eventHander
  };
};

export const ListCardDelete: ListActionCreator = (eventHander: ListEventHandler) => {
  return {
    label: 'Delete',
    type: CardActionType.delete,
    handler: eventHander
  };
};
