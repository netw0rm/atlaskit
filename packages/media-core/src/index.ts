import {MediaItemType} from './item';

export * from './item';
export * from './actions';
export * from './config';
export * from './mediaTypes';

export interface RemoteCollection {
  id: string;
  apiConfig: {
    clientId: string;
    token: string;
    baseURL: string;
  };
}

export interface MediaCollectionItem {
  id: string;
  mediaItemType: MediaItemType;
}

export interface MediaCollection {
  id: string;
  items: Array<MediaCollectionItem>;
}

export interface Metadata {
  id: string;
  mediaType: string;
  mimeType: string;
  name: string;
  processingStatus: string;
  size: number;
}

export default {};
