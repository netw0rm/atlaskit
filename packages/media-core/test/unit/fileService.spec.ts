import * as sinon from 'sinon';

import {MediaFileService} from '../../src/services/fileService';
import {JwtTokenProvider, MediaItem} from '../../src';
import {LRUCache} from 'lru-fast';
import {FileDetails} from '../../src/item';

const serviceHost = 'some-host';
const token = 'some-token';

const fileId = 'some-file-id';
const unprocessedFileId = 'some-unprocessed-file-id';
const clientId = 'some-client-id';
const collection = 'some-collection';
const defaultFileDetails = {
  id: 'some-file-id',
  mediaType: 'image',
  mimeType: 'some-mime-type',
  name: 'some-name',
  processingStatus: 'succeeded',
  size: 12345,
  artifacts: {
    'document.pdf': {href: `/file/${fileId}/artifact/document.pdf`},
    'presentation.ppt': {href: `/file/${fileId}/artifact/presentation.ppt`}
  }
};

describe('MediaFileService', () => {
  let tokenProvider: JwtTokenProvider;
  let fileService: MediaFileService;

  let xhr: any;
  let requests: Array<any>;

  const setupFakeXhr = () => {
    xhr = sinon.useFakeXMLHttpRequest();
    requests = [];

    xhr.onCreate = function (xhr: any) {
      requests.push(xhr);
    };
  };

  const respondFakeXhr = (fileDetails?: FileDetails) => {
    setTimeout(() => {
      const mockedResponse = {
        data: fileDetails || defaultFileDetails
      };
      if (requests[0]) {
        requests[0].respond(200, { 'Content-Type': 'application/json' }, JSON.stringify(mockedResponse));
      }
    });
  };

  beforeEach(() => {
    setupFakeXhr();
    tokenProvider = jest.fn(() => Promise.resolve(token));
    const cache = new LRUCache<string, MediaItem>(0);
    fileService = new MediaFileService({ serviceHost, tokenProvider }, cache);
  });

  afterEach(function () {
    xhr.restore();
  });

  it('should resolve file item from collection given or not', () => {
    const response = fileService.getFileItem(fileId, clientId, collection)
      .then(fileItem => {
        expect(fileItem.type).toBe('file');
        expect(fileItem.details).toEqual(defaultFileDetails);
      })
      .then(() => {
        // Validate call to token provider
        expect(tokenProvider).toHaveBeenCalledWith(collection);
      })
      .then(() => {
        expect(requests[0].url).toMatchSnapshot();
      });

    respondFakeXhr();

    return response;
  });

  it('should resolve file item from collection given', () => {
    const response = fileService.getFileItem(fileId, clientId)
      .then(fileItem => {
        expect(fileItem.type).toBe('file');
        expect(fileItem.details).toEqual(defaultFileDetails);
      })
      .then(() => {
        // Validate call to token provider
        expect(tokenProvider).toHaveBeenCalledWith(undefined);
      })
      .then(() => {
        expect(requests[0].url).toMatchSnapshot();
      });

    respondFakeXhr();

    return response;
  });

  it('should reject server responded with 500', () => {
    const response = fileService.getFileItem('some-dodgy-file-id', clientId, collection)
      .then(
        () => { throw new Error('The function getFileItem should fail'); },
        (error) => expect(error).toBeDefined()
      );

    setTimeout(() => { requests[0].respond(500, { }, ''); });
    return response;
  });

  describe('cache', () => {
    const shouldReturnFileFromService = (id: string, cache: LRUCache<string, MediaItem>, fileDetails?: FileDetails) => {
      tokenProvider = jest.fn(() => Promise.resolve(token));
      fileService = new MediaFileService({ serviceHost, tokenProvider }, cache);
      const response = fileService.getFileItem(id, clientId, collection).then(() => {
        expect(tokenProvider).toHaveBeenCalledTimes(1);
      });

      respondFakeXhr(fileDetails);

      return response;
    };

    const shouldReturnFileFromCache = (id: string, cache: LRUCache<string, MediaItem>) => {
      tokenProvider = jest.fn(() => Promise.resolve(token));
      fileService = new MediaFileService({ serviceHost, tokenProvider }, cache);
      return fileService.getFileItem(id, clientId, collection).then(() => {
        expect(tokenProvider).not.toHaveBeenCalled();
      });
    };

    it('should cache processed files', () => {
      const cache = new LRUCache<string, MediaItem>(1);
      return shouldReturnFileFromService(fileId, cache).then(() => shouldReturnFileFromCache(fileId, cache));
    });

    it('should not cache processed files if caching is disabled', () => {
      const cache = new LRUCache<string, MediaItem>(0);
      return shouldReturnFileFromService(fileId, cache).then(() => {
        xhr.restore();
        setupFakeXhr();
        shouldReturnFileFromService(fileId, cache);
      });
    });

    it('should not cache unprocessed files', () => {
      const unprocessedFileDetails: FileDetails = {
        id: 'some-file-id',
        mediaType: 'image',
        mimeType: 'some-mime-type',
        name: 'some-name',
        processingStatus: 'pending',
        size: 12345,
        artifacts: {
          'document.pdf': { href: `/file/${fileId}/artifact/document.pdf` },
          'presentation.ppt': { href: `/file/${fileId}/artifact/presentation.ppt` }
        }
      };
      const cache = new LRUCache<string, MediaItem>(1);
      return shouldReturnFileFromService(unprocessedFileId, cache, unprocessedFileDetails).then(() => {
        xhr.restore();
        setupFakeXhr();
        shouldReturnFileFromService(unprocessedFileId, cache, unprocessedFileDetails);
      });
    });
  });
});
