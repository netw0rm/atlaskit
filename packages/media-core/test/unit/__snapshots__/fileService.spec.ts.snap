// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaFileService should resolve file item from collection given 1`] = `"some-host/file/some-file-id?token=some-token&client=some-client-id"`;

exports[`MediaFileService should resolve file item from collection given or not 1`] = `"some-host/file/some-file-id?collection=some-collection&token=some-token&client=some-client-id"`;
